### PR TITLE
bugfix: parent workflow, when signaling child workflow, can experienc…

### PR DESCRIPTION
**Pull request by wxing1292**
…e deadlock, if child workflow, at the same time, is completed.

### CODE REVIEW

File: service/history/transferQueueProcessor.go [hunk 1]
```diff
@@ -328,26 +328,26 @@ func (t *transferQueueProcessorImpl) processActivityTask(task *persistence.Trans
 	if err != nil {
 		return err
 	}
+	defer release()
```
Comment (samarabbas) [inline service/history/transferQueueProcessor.go]: Can you immediate call 'defer release()' after acquiring the release object?

File: service/history/transferQueueProcessor.go [hunk 2]
```diff
@@ -328,26 +328,26 @@ func (t *transferQueueProcessorImpl) processActivityTask(task *persistence.Trans
...
 
-	if ai, found := mb.GetActivityInfo(task.ScheduleID); found {
+	if ai, found := msBuilder.GetActivityInfo(task.ScheduleID); found {
 		timeout = ai.ScheduleToStartTimeout
 	} else {
 		logging.LogDuplicateTransferTaskEvent(t.logger, persistence.TransferTaskTypeActivityTask, task.TaskID, task.ScheduleID)
 	}
-	release()
 
+	release()
```
Comment (samarabbas) [inline service/history/transferQueueProcessor.go]: Can you add a comment on why calling the release here is important.

File: service/history/transferQueueProcessor.go [hunk 3]
```diff
@@ -455,16 +456,30 @@ func (t *transferQueueProcessorImpl) processCloseExecution(task *persistence.Tra
...
+	completionEvent, _ := msBuilder.GetCompletionEvent()
+	parentDomainID := msBuilder.executionInfo.ParentDomainID
+	parentWorkflowID := msBuilder.executionInfo.ParentWorkflowID
+	parentRunID := msBuilder.executionInfo.ParentRunID
+	initiatedID := msBuilder.executionInfo.InitiatedID
+
+	workflowTypeName := msBuilder.executionInfo.WorkflowTypeName
+	workflowStartTimestamp := msBuilder.executionInfo.StartTimestamp.UnixNano()
+	workflowCloseTimestamp := msBuilder.getLastUpdatedTimestamp()
+	workflowStatue := getWorkflowExecutionCloseStatus(msBuilder.executionInfo.CloseStatus)
```
Comment (samarabbas) [inline service/history/transferQueueProcessor.go]: nit: typo

File: service/history/transferQueueProcessor.go [hunk 4]
```diff
@@ -455,16 +456,30 @@ func (t *transferQueueProcessorImpl) processCloseExecution(task *persistence.Tra
 		return err
 	}
 
+	isChildWorkflow := msBuilder.hasParentExecution()
+	workflowCloseStatus := msBuilder.executionInfo.CloseStatus
+	completionEvent, _ := msBuilder.GetCompletionEvent()
```
Comment (samarabbas) [inline service/history/transferQueueProcessor.go]: We will unnecessarily deserialize the completionEvent even for execution which does not have a Parent.  Can you bring this back with the If check?

File: service/history/transferQueueProcessor.go (hunk 1 above)
Comment (wxing1292) [reply service/history/transferQueueProcessor.go]: there can be cases when cache is full and context, release being nil, while err not being nil, doing defer release() before error checking can lead to null pointer issue

File: service/history/transferQueueProcessor.go (hunk 2 above)
Comment (wxing1292) [reply service/history/transferQueueProcessor.go]: sure

File: service/history/transferQueueProcessor.go (hunk 3 above)
Comment (wxing1292) [reply service/history/transferQueueProcessor.go]: fixed

File: service/history/transferQueueProcessor.go (hunk 4 above)
Comment (wxing1292) [reply service/history/transferQueueProcessor.go]: sure

**Pull request merged**